tools/circt-lec/Solver.h

```diff
@@ -0,0 +1,75 @@
+//===-- Solver.h - SMT solver interface -------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// This file defines a SMT solver interface for the `circt-lec` tool.
+///
+//===----------------------------------------------------------------------===//
+
+// NOLINTNEXTLINE
+#ifndef TOOLS_CIRCT_LEC_SOLVER_H
+#define TOOLS_CIRCT_LEC_SOLVER_H
+
+#include "circt/Support/LLVM.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/Value.h"
+#include <z3++.h>
+
+/// A satisfiability checker for circuit equivalence
+///
+/// This class interfaces with an external SMT solver acting as a logical
+/// engine. First spawn two circuits through `addCircuit`; after collecting
+/// their logical constraints, the `solve` method will compare them and report
+/// whether they result to be equivalent or, when not, also printing a model
+/// acting as a counterexample.
+class Solver {
+public:
+  Solver(mlir::MLIRContext *mlirCtx, bool statisticsOpt);
+  ~Solver() = default;
+
+  /// Solve the equivalence problem between the two circuits, then present the
+  /// results to the user.
+  mlir::LogicalResult solve();
+
+  class Circuit;
+  /// Create a new circuit to be compared and return it.
+  Circuit *addCircuit(llvm::StringRef name);
```

Comment:
After this PR lands, we could look into turning this into something like `getCircuitA` and `getCircuitB`, since the solver has this `Circuit *circuit[2]` member that really forces two circuits to be present. Having `addCircuit` makes it easy to abuse the solver by not having exactly two calls to `addCircuit`. Just some random thoughts on future refactorings.